perf/benchmarkapps/QpsWorker/Infrastructure/ClientRunner.cs

```diff
@@ -0,0 +1,352 @@
+                TimeSystem = timeSnapshot.PrivilegedProcessorTime.TotalSeconds
+            };
+        }
+
+        public async Task StopAsync()
+        {
+            _stoppedCts.Cancel();
+            foreach (var runnerTask in _runnerTasks)
+            {
+                await runnerTask;
+            }
+            foreach (var channel in _channels)
+            {
+                await channel.ShutdownAsync();
+            }
+        }
+
+        private void RunUnary(GrpcChannel channel, IInterarrivalTimer timer)
+        {
+            var client = new BenchmarkService.BenchmarkServiceClient(channel);
+            var request = CreateSimpleRequest();
+            var stopwatch = new Stopwatch();
+
+            while (!_stoppedCts.Token.IsCancellationRequested)
+            {
+                stopwatch.Restart();
+                client.UnaryCall(request);
+                stopwatch.Stop();
+
+                // spec requires data point in nanoseconds.
+                _threadLocalHistogram.Value!.AddObservation(stopwatch.Elapsed.TotalSeconds * SecondsToNanos);
+
+                timer.WaitForNext();
+            }
+        }
+
+        private async Task RunUnaryAsync(GrpcChannel channel, IInterarrivalTimer timer)
+        {
+            var client = new BenchmarkService.BenchmarkServiceClient(channel);
+            var request = CreateSimpleRequest();
+            var stopwatch = new Stopwatch();
+
+            while (!_stoppedCts.Token.IsCancellationRequested)
+            {
+                stopwatch.Restart();
+                await client.UnaryCallAsync(request);
+                stopwatch.Stop();
+
+                // spec requires data point in nanoseconds.
+                _threadLocalHistogram.Value!.AddObservation(stopwatch.Elapsed.TotalSeconds * SecondsToNanos);
+
+                await timer.WaitForNextAsync();
+            }
+        }
+
+        private async Task RunStreamingPingPongAsync(GrpcChannel channel, IInterarrivalTimer timer)
+        {
+            var client = new BenchmarkService.BenchmarkServiceClient(channel);
+            var request = CreateSimpleRequest();
+            var stopwatch = new Stopwatch();
+
+            using (var call = client.StreamingCall())
+            {
+                while (!_stoppedCts.Token.IsCancellationRequested)
+                {
+                    stopwatch.Restart();
+                    await call.RequestStream.WriteAsync(request);
+                    await call.ResponseStream.MoveNext();
+                    stopwatch.Stop();
+
+                    // spec requires data point in nanoseconds.
+                    _threadLocalHistogram.Value!.AddObservation(stopwatch.Elapsed.TotalSeconds * SecondsToNanos);
+
+                    await timer.WaitForNextAsync();
+                }
+
+                // finish the streaming call
+                await call.RequestStream.CompleteAsync();
+                if (!await call.ResponseStream.MoveNext())
```

Comment:
Ah, ok, I think for streaming you have the condition reversed.

call.RequestStream.CompleteAsync() half closes the request stream after which the server should not send any more responses (but you're checking the opposite).
Original logic: https://github.com/grpc/grpc/blob/b639e09c90917b39418df3e13a326cbcea0bfbd5/src/csharp/Grpc.IntegrationTesting/ClientRunners.cs#L275